ooasp/run.py

```diff
@@ -1,314 +1,591 @@
+                        action="store_true",
+                        help="Verbose output")
+    parser.add_argument(
+        "--view",
+        action="store_true",
+        help="Saves the solution as a PNG generated by clingraph",
+    )
+    # ------------------------Smart generation options ------------------------
+    parser.add_argument("--timeout", type=int, help="Timeout in seconds")
+    parser.add_argument(
+        "--smart-generation",
+        type=str,
+        help=f"""Specifies what functions for smart generation to use.
+                    Possible options: {list(SMART_FUNCTIONS.keys())}.
+                    Input has to be comma-separated list.
+                    Functions are executed in the order specified.""",
+    )
+    # ------------------------Initial objects------------------------
+    for c in CLASSES:
+        parser.add_argument(f"--{c}",
+                            type=int,
+                            default=0,
+                            help=f"Number of {c}s")
+    return parser
 
-def ground(ctl, size, o):
-    log(f"\t\tGrounding {size} {o}")
-    ctl.ground([("domain", [Number(size), Function(o, [])])])
-    if size > 0:
-        ctl.release_external(Function("active", [Number(size - 1)]))
-    ctl.assign_external(Function("active", [Number(size)]), True)
 
+class OOASPRacksSolver:
+    """
+    Solver for the OOASP using the racks problem as example.
+    The approach is generalizable to other problems.
+    """
 
-def add_object(ctl, o, size, association=None):
-    log(f"\t\tAdding object  {o},{size}")
-    obj_atom = f"ooasp_isa({o},{size})"
-    ctl.add("domain", [str(size), "object"], f"user({obj_atom}).")
-    config.append(obj_atom)
-    config.append(f"user({obj_atom})")
-    if association is not None:
+    def __init__(
+        self,
+        initial_objects=None,
+        smart_generation_functions=None,
+        verbose=False,
+        view=False,
+        timeout=None,
+    ):
+        """
+        Initialize the solver.
+
+        Args:
+            initial_objects (list): List of initial objects to start with.
+            smart_generation_functions (list): List of smart generation functions to use for incremental generation.
+            verbose (bool): Verbose output will show the logs
+            view (bool): If True, saves the solution as a PNG generated by clingraph
+            timeout (int): Timeout in seconds
+        """
+        self.initial_objects = initial_objects if initial_objects is not None else []
+        self.smart_generation_functions = (smart_generation_functions
+                                           if smart_generation_functions
+                                           is not None else [])
+        self.view = view
+        self.timeout = timeout if timeout is not None else 1200
+
+        self.ctl = Control([
+            "1",
+            "--warn=none",
+            "-c config_name=c1",
+            "-c kb_name=k1",
+            "-t3",
```

Comment:
What's `t3`?

Reply:
Instruction for clingo to use 3 threads, the verbose command would be `--parallel-mode=3`. Do you prefer that?